### PR TITLE
feat: add options to pass custom regex to conventional-commits-parser

### LINKED ISSUE #175: Feature request: Custom regex instead of `conventional-commits-parser`
This action is really great and covers most of the cases we need, but unfortunately we have slightly different format than "conventional commits". (Our format looks like `[ADD] New feature`)

To solve this, it would be great to have option to specify custom regex that validates the whole title, completely replacing the conventional commits parser

### PR DESCRIPTION
**Pull request by Jomshir98**
Resolves #175 

The PR I used for testing:
[Jomshir98/action-semantic-pull-request#1](https://github.com/Jomshir98/action-semantic-pull-request/pull/1)

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -65,6 +65,13 @@ The action works without configuration, however you can provide options for cust
             The subject "{subject}" found in the pull request title "{title}"
             didn't match the configured pattern. Please ensure that the subject
             doesn't start with an uppercase character.
+          # Configure custom regex that will be passed to conventional-commits-parser.
+          # This can be used to specify different required format of the title.
+          # The correspondence is used to define what capturing group of headerPattern captures what header part.
+          # The example values are the default.
+          # For more details see: https://github.com/conventional-changelog/conventional-changelog/tree/master/packages/conventional-commits-parser#headerpattern
```
Comment (amannn) [inline README.md:72]: I tried to rephrase it a bit:

```suggestion
          # If you're using a format for the PR title that differs from the traditional Conventional
          # Commits spec, you can use these options to customize the parsing of the type, scope and
          # subject. The `headerPattern` should contain a regex where the capturing groups in parentheses
          # correspond to the parts listed in `headerPatternCorrespondence`.
          # See: https://github.com/conventional-changelog/conventional-changelog/tree/master/packages/conventional-commits-parser#headerpattern
```

What do you think?

Also, would you mind moving this and the `wip` option to the bottom of the config list? I think they might be less commonly used than the other ones.

File: action.yml [hunk 2]
```diff
@@ -23,6 +23,12 @@ inputs:
   subjectPatternError:
     description: "If `subjectPattern` is configured, you can use this property to override the default error message that is shown when the pattern doesn't match. The variables `subject` and `title` can be used within the message."
     required: false
+  headerPattern:
+    description: "Configure custom regex that will be passed to conventional-commits-parser. For more details see: https://github.com/conventional-changelog/conventional-changelog/tree/master/packages/conventional-commits-parser#headerpattern"
```
Comment (amannn) [inline action.yml:27]:
```suggestion
    description: "If you're using a format for the PR title that differs from the traditional Conventional Commits spec, you can use this to customize the parsing of the type, scope and subject. The `headerPattern` should contain a regex where the capturing groups in parentheses correspond to the parts listed in `headerPatternCorrespondence`. For more details see: https://github.com/conventional-changelog/conventional-changelog/tree/master/packages/conventional-commits-parser#headerpattern"
```

File: action.yml [hunk 3]
```diff
@@ -23,6 +23,12 @@ inputs:
...
+    required: false
+  headerPatternCorrespondence:
+    description: "If `headerPattern` is configured, you can use this property to also configure which RegEx groups match what part of the semantic commit."
```
Comment (amannn) [inline action.yml:30]:
```suggestion
    description: "If `headerPattern` is configured, you can use this to define which capturing groups correspond to the type, scope and subject."
```

**Review by amannn — COMMENTED** [review]
Thank you very much for your contribution! The code looks excellent, I only have some minor suggestions for the descriptions in the docs.

What do you think?

**Comment by Jomshir98** [comment]
@amannn 
I fully agree with the points in the review; writing documentation never really was my strong side.
As for the options, I feel the GitHub enterprise will be the least used one, but of course I don't have any stats about this and personally when I am reading docs, I usually read all of them.

**Review by amannn — APPROVED** [review]
Perfect, thank you so much! 🙌

> As for the options, I feel the GitHub enterprise will be the least used one, but of course I don't have any stats about this and personally when I am reading docs, I usually read all of them.

That could be, yes – but I think we can leave it for now.

**Pull request merged**